src/backend/distributed/executor/multi_utility.c

```diff
@@ -2921,6 +2934,126 @@ ErrorIfUnsupportedTruncateStmt(TruncateStmt *truncateStatement)
 }
 
 
+/*
+ * ProcessTruncateStatement handles distributed locking
+ * of truncated tables before standard utility takes over.
+ *
+ * This is only for distributed serialization of truncate commands.
+ * The function assumes that there is no foreign key relation between
+ * non-distributed and distributed relations.
+ */
+static void
+ProcessTruncateStatement(TruncateStmt *truncateStatement)
+{
+	List *distributedRelationList = NIL;
+
+	ListCell *relationCell = NULL;
+	foreach(relationCell, truncateStatement->relations)
+	{
+		RangeVar *relationRV = (RangeVar *) lfirst(relationCell);
+		Relation relation = heap_openrv(relationRV, NoLock);
+		Oid relationId = RelationGetRelid(relation);
+		DistTableCacheEntry *cacheEntry = NULL;
+		List *referencingTableList = NIL;
+		ListCell *referencingTableCell = NULL;
+
+		if (!IsDistributedTable(relationId))
+		{
+			heap_close(relation, NoLock);
+			continue;
+		}
+
+		if (list_member_oid(distributedRelationList, relationId))
+		{
+			heap_close(relation, NoLock);
+			continue;
+		}
+
+		distributedRelationList = lappend_oid(distributedRelationList, relationId);
+
+		cacheEntry = DistributedTableCacheEntry(relationId);
+		Assert(cacheEntry != NULL);
+
+		referencingTableList = cacheEntry->referencingRelationsViaForeignKey;
+		foreach(referencingTableCell, referencingTableList)
+		{
+			Oid referencingRelationId = lfirst_oid(referencingTableCell);
+			distributedRelationList = list_append_unique_oid(distributedRelationList,
+															 referencingRelationId);
+		}
+
+		heap_close(relation, NoLock);
+	}
+
+	if (distributedRelationList != NIL)
+	{
+		BeginOrContinueCoordinatedTransaction();
+		AcquireDistributedLockOnRelations(distributedRelationList, AccessExclusiveLock,
```

Comment:
We've a function named `ClusterHasKnownMetadataWorkers()` which returns true if the cluster is an MX cluster.

So, we should not be getting into this logic at all if `!ClusterHasKnownMetadataWorkers()`. Otherwise, we'd be sending these commands even if we're not in an MX cluster.

Reply:
correct, I was only checking `ShouldSyncTableMetadata()` per relation. This is a shortcut to not  to enter that path.